Feature/notifications

micahmo opened the pull request:
## Pull Request Description

This PR introduces OS notifications for inbox messages. All message checking and notification generation is done locally (i.e., no Firebase needed).

**New Packages**
 - [`flutter_local_notifications`](https://pub.dev/packages/flutter_local_notifications)
   - Has the ability to generate local notifications, manage permissions, etc.
 - [`background_fetch`](https://pub.dev/packages/background_fetch)
   - Allows the app to periodically run a process in the background (in this case, polling for new messages), even when closed.

**Notes**
> The following is a list of caveats and/or intentional limitations for this first round of implementation. Most of these things can be improved in the future, but this first round works great as is!
- I only set this up for Android. Perhaps @hjiangsu could do the iOS setup at some point, but it requires a bit of work for both packages. ([flutter_local_notifications setup](https://pub.dev/packages/flutter_local_notifications#-ios-setup) and [background_fetch setup](https://pub.dev/packages/background_fetch)).
- Enabling notifications will most likely increase battery usage due to the background checks. This is noted in the new setting description.
- There are some ["Release build configuration"](https://pub.dev/packages/flutter_local_notifications#release-build-configuration) changes suggested by `flutter_local_notifications` that I did not go through. I'd be curious to know if it works as-is (since it works fine in Debug), but we can test together.
- Clicking a notification only brings you to inbox page. It does not specifically focus the relevant message.
  - My idea to fix this in a future iteration is to add a new page which can display a single message and be shown on top of the navigation stack. This allows you to interact with the message, then go back to whatever you were doing in Thunder. (Right now, the page scroll to the Inbox page doesn't work so well if you're looking at a post, for example.)
- Notifications only work for replies, not mentions or messages.
- Notifications only work for the active account in Thunder.
- The interval of `background_fetch` is not super configurable (at all on iOS, and not without more permissions on Android). But I think the default of 15(-ish) minutes is fine!
   - For the sake of the demo video, the polling time was reduced. The code includes comments on how to do this.

## Issue Being Fixed

<!-- Please describe the problem that is being fixed and, if applicable, reference a GitHub issue -->

Issue Number: First step / partial implementation of #125 and #219.

## Screenshots / Recordings

<!-- This section is optional but highly recommended to show off your changes! -->

### Demo of permissions flow

https://github.com/thunder-app/thunder/assets/7417301/57885668-7094-4a81-b27c-fef283f8a5e2

### Demo of receiving notifications while app is open

https://github.com/thunder-app/thunder/assets/7417301/f9d29af6-e0cf-4d35-b57f-f38efde20cda

### Demo of receiving notifications while app is closed

https://github.com/thunder-app/thunder/assets/7417301/825d5e6c-ad0a-46e0-9858-c9d63c3ff6d6

## Checklist

- [ ] Did you update CHANGELOG.md?
- [ ] Did you use localized strings where applicable?
- [ ] Did you add `semanticLabel`s where applicable for accessibility?

hjiangsu commented:
I would just like to say that this is awesome! I was also looking at ways to have notifications without having to host a server holding JWT tokens (for security purposes)

I'll definitely take a look at this when I have time and see if it can be adapted to iOS as well 😄

micahmo commented:
> I would just like to say that this is awesome! I was also looking at ways to have notifications without having to host a server holding JWT tokens (for security purposes)

Thanks, yeah, this seems to be a good enough solution, as long as we're ok with Thunder making the occasional API call!

If you'd like to try this out on your Android emulator to get a feel for how it works, I'd recommend a couple of changes to make testing easier.

* Change this value to `1`.
  https://github.com/thunder-app/thunder/blob/04cd9d7502551e0618e62e6bc84bc2adb0df26a5/lib/main.dart#L237
* Uncomment this line.
  https://github.com/thunder-app/thunder/blob/04cd9d7502551e0618e62e6bc84bc2adb0df26a5/lib/main.dart#L247
* Comment out the `where` part of this line.
  https://github.com/thunder-app/thunder/blob/04cd9d7502551e0618e62e6bc84bc2adb0df26a5/lib/utils/notifications.dart#L57
* Mark some messages as unread _or_ change this to `false`.
  https://github.com/thunder-app/thunder/blob/04cd9d7502551e0618e62e6bc84bc2adb0df26a5/lib/utils/notifications.dart#L49

With those things set, you should get some notifications for some/all inbox messages about once per minute whether the app is running or not!

hjiangsu commented:
So I've been looking into the packages, especially [background_fetch](https://pub.dev/packages/background_fetch), and it seems like there might be some inconsistencies in how this works (depending on platform and manufacturer)

For iOS, background fetching will not work when the app is terminated. This means that we can't really use this package reliably to fetch for notifications every ~15 mins. This will only work if the app is either in the foreground, or when its not terminated (in the background)

For Android, it seems like the reliability of background fetching varies depending on the device manufacturer (https://dontkillmyapp.com/). There seems to be some workarounds for the specific manufacturers but each one is different.
- https://github.com/transistorsoft/flutter_background_fetch/issues/355
- https://github.com/transistorsoft/flutter_background_fetch/issues/339

hjiangsu commented:
Yeah, unfortunately, I think the only viable solution on iOS would be to set up a server to handle push notifications. I could play around with it a bit and see how that would work, but needing to have the JWT token stored on the server in order to fetch notifications feels like a big security risk.

There might be a potential for self-hosting the server which reduces the security risk for those who are able to do so, but I don't know how that interacts with the push notification services (whether or not that's even possible)

I think we can still go ahead and push this local notifications since its really useful to have, but we should be clear that it's experimental and might not work on all Android devices (and provide some instructions on what to do to prevent the app from being killed in the background)

micahmo commented:
> we should be clear that it's experimental and might not work on all Android devices (and provide some instructions on what to do to prevent the app from being killed in the background)

Any recommendations for how we should communicate this? Maybe a popup warning when enabling the feature?

hjiangsu commented:
I think adding "Experimental" to the subheading of the option works well to let the user know that it may not fully work. We can also have a popup dialog which explains how the notifications work so that its more transparent
- Only happens on ~15 mins intervals, disable battery optimizations to give it the best chance to work properly, etc.

micahmo commented:
> I think adding "Experimental" to the subheading of the option works well to let the user know that it may not fully work. We can also have a popup dialog which explains how the notifications work so that its more transparent

Sounds good! How does this look?

https://github.com/thunder-app/thunder/assets/7417301/851e6302-5a75-47a9-88a4-21eab7a0ffad

Here's a screenshot of just the dialog so you can read it.

| ![image](https://github.com/thunder-app/thunder/assets/7417301/88f1ffd9-77eb-47b4-b276-b94f42c8a2c8) |
| - |

micahmo commented on lib/settings/pages/general_settings_page.dart:
```diff
@@ -662,7 +702,7 @@ class _GeneralSettingsPageState extends State<GeneralSettingsPage> with SingleTi
                   subtitle: enableInboxNotifications
                       ? !kIsWeb && Platform.isAndroid && areAndroidNotificationsAllowed == true
-                          ? l10n.backgroundCheckWarning
+                          ? null
```
I took this out because this information is now covered in the dialog (and the experimental heading) and to keep the settings page more clean.

micahmo commented:
> It'll take some time for me to review this PR

No problem at all! I really appreciate that you're willing to review and accept this feature. Take all the time you need.

> I'll most likely just generate a release build for you to test

Sounds good!

hjiangsu commented on android/build.gradle:
```diff
@@ -1,12 +1,20 @@
+    ext {
+        compileSdkVersion   = 33
+        targetSdkVersion    = 33
+        appCompatVersion    = "1.6.1"
```
Just curious, but where did "1.6.1" come from?

micahmo replied:
It essentially comes from [this section](https://github.com/transistorsoft/flutter_background_fetch/blob/master/help/INSTALL-ANDROID.md#androidbuildgradle) of the `background_fetch` setup guide.

``` diff
+       appCompatVersion    = "1.4.2"           // or latest
```

Notice it allows for the latest. I believe I ran into an issue and someone suggesting upgrading, so I looked up the latest  [here](https://developer.android.com/jetpack/androidx/releases/appcompat), and it is 1.6.1.

Sorry I don't remember the specific issue that prompted me to upgrade.

hjiangsu replied:
That's all good! I was just curious

hjiangsu commented on lib/settings/pages/general_settings_page.dart:
```diff
@@ -221,13 +236,24 @@ class _GeneralSettingsPageState extends State<GeneralSettingsPage> with SingleTi
+    WidgetsBinding.instance.addPostFrameCallback((_) async {
+      _initPreferences();
+
+      // Check whether Android is currently allowing Thunder to send notifications
```
I would place all this logic into its own function (e.g., similar to `_initPreferences`). This way, you can perform the async operations!

hjiangsu replied:
Ahh, I think I misinterpreted the comment. We can do something like this so that it doesn't mutate `areAndroidNotificationsAllowed` directly:

Although in this case, it might be fine!

```dart
bool _areAndroidNotificationsAllowed = await androidFlutterLocalNotificationsPlugin?.areNotificationsEnabled();

setState(() {
  areAndroidNotificationsAllowed: _areAndroidNotificationsAllowed,
});
```

micahmo replied:
Ok I updated it to do this! I think the end result is the same (the UI is redrawn with the current state), but I agree that this provides a clearer intention as to why `setState` is being called at all.

hjiangsu reviewed (commented):
I've pushed some commits to update with the latest develop changes, and also to add in release build configuration! I can create a release build for you to test out if you'd like (I can send it over matrix)

I also did a quick code review and had some comments!

micahmo commented on android/app/build.gradle:
```diff
@@ -52,9 +55,12 @@ android {
         versionCode flutterVersionCode.toInteger()
         versionName flutterVersionName
+        // The following line is required by flutter_local_notifications
+        multiDexEnabled true
```
Good catch! I'll take it out and we'll see if things still work. 😊

micahmo commented on android/app/src/main/AndroidManifest.xml:
I don't believe so! If you check the first bullet point under the heading you linked, it says...

> To schedule notifications the following changes are needed

We are not using the schedule functionality, so I don't believe anything there applies to us.

P.S. At first I was hoping that the schedule functionality allowed you to execute some code and perhaps conditionally generate a notification, but unfortunately, it only lets you schedule a predefined notification at a predefined interval, so it wouldn't be much help.

hjiangsu replied:
Sounds good! Just double-checking 😄

micahmo commented:
> I've pushed some commits to update with the latest develop changes, and also to add in release build configuration! I can create a release build for you to test out if you'd like (I can send it over matrix)
> 
> I also did a quick code review and had some comments!

Thank you so much for reviewing and adding the release configs! I've addressed your PR comments. If/when you get a chance, feel free to shoot over a release build in Matrix and I'll see how things work!
